### PR TITLE
PR :Add watchdog dependency and watcher on qrc processing when qss changes

### PR DESCRIPTION
**Pull request by goanpeca**
- [x] modify scripts, so they can run from outside the scripts folder
- [x] Added an Observer using watchdog so that the qrc files are generated automatically if changes are detected on the QSS.

Works like this:

```bash
python process_qrc.py --watch
```

### CODE REVIEW

**Comment by goanpeca** [comment]
I made this to make life easier, when trying to debug the ComboBox Issue on Spyder. However I hit an block in that I am not being able to refresh the stylesheets on Spyder without a restart. I am working on develop mode

```bash
python setup.py develop
```

And was expecting that a reload would yield a new CSS but I keep getting the old one :-\

Thoughts @ccordoba12 @dpizetta ?

File: script/process_qrc.py [hunk 1]
```diff
@@ -23,31 +23,76 @@
...
+# Third party imports
+from watchdog.events import FileSystemEventHandler
+from watchdog.observers import Observer
+
+# Constants
+HERE = os.path.abspath(os.path.dirname(__file__))
+REPO_ROOT = os.path.dirname(HERE)
+
+
+class QSSFileHandler(FileSystemEventHandler):
```
Comment (goanpeca) [inline script/process_qrc.py]: This watcher (with a small tweak) could also be applied to other scripts.

Comment (dpizetta) [reply script/process_qrc.py]: Nice! Tks

**Comment by dpizetta** [comment]
>And was expecting that a reload would yield a new CSS but I keep getting the old one :-\

You confirmed that the watcher run the processing of qss? Every time it runs you should get a message on the same terminal you started the watcher, informing the changes and compilation messages to provide feedback.

Changing the style.qss file triggers the watcher, but it lists these files:
```

<class 'watchdog.events.DirModifiedEvent'>
<class 'watchdog.events.FileModifiedEvent'>
/home/pizetta/workspace/qdark-combo/qdarkstyle/.goutputstream-OGFA1Z False
<class 'watchdog.events.FileModifiedEvent'>
/home/pizetta/workspace/qdark-combo/qdarkstyle/.goutputstream-OGFA1Z False
<class 'watchdog.events.FileModifiedEvent'>
/home/pizetta/workspace/qdark-combo/qdarkstyle/.goutputstream-OGFA1Z False
<class 'watchdog.events.FileModifiedEvent'>
/home/pizetta/workspace/qdark-combo/qdarkstyle/.goutputstream-OGFA1Z False
<class 'watchdog.events.FileModifiedEvent'>
/home/pizetta/workspace/qdark-combo/qdarkstyle/.goutputstream-OGFA1Z False
<class 'watchdog.events.FileModifiedEvent'>
/home/pizetta/workspace/qdark-combo/qdarkstyle/.goutputstream-OGFA1Z False
<class 'watchdog.events.FileModifiedEvent'>
/home/pizetta/workspace/qdark-combo/qdarkstyle/.goutputstream-OGFA1Z False
<class 'watchdog.events.DirModifiedEvent'>

```

None of them is qss. So, the IF testing the path will not work there.

**Comment by goanpeca** [comment]
>You confirmed that the watcher run the processing of qss? Every time it runs you should get a message on the same terminal you started the watcher, informing the changes and compilation messages to provide feedback.

Yes, it seems to be a limitation of loading resources via import, instead of the direct use of QSS file.

>None of them is qss. So, the IF testing the path will not work there.

How did you ran and where did you run it?

**Comment by dpizetta** [comment]
I have not tried to change the imported file with running app. Actually even after terminating the app you may need to clean cache files. I had this problem with other projects. You can use the reload(), a python function that refreshes the import at runtime.

The second point. I run using your implementation making python process_qrc.py --watch inside script folder. I print the path out of IF statement, before it. Also, I print type() to confirm. Your IF is getting always False. I don't know why but it not list style.qss has changed, because I changed it manually on editor. Even more strange is that all the files in the folder (...rc.py) have their date update...but they are not changed, because the run_process() is not called.  :0

**Comment by goanpeca** [comment]
>You can use the reload(), a python function that refreshes the import at runtime.

That did not work either, I tried that

>I have not tried to change the imported file with running app. Actually even after terminating the app you may need to clean cache files. I had this problem with other projects. You can use the reload(), a python function that refreshes the import at runtime.

I have ran projects with Qt with custom CSS that is refreshed on runtime and the application refreshed successfully  

>The second point. I run using your implementation making python process_qrc.py --watch inside script folder. I print the path out of IF statement, before it. Also, I print type() to confirm. Your IF is getting always False. I don't know why but it not list style.qss has changed, because I changed it manually on editor. Even more strange is that all the files in the folder (...rc.py) have their date update...but they are not changed, because the run_process() is not called. :0

That is weird indeed, will look deeper into it.

File: setup.py [hunk 2]
```diff
@@ -32,5 +32,8 @@
         'Topic :: Software Development :: Libraries :: Application Frameworks'
     ],
     zip_safe=False,  # don't use eggs
-    entry_points={"console_scripts": ["qdarkstyle=qdarkstyle.__main__:main"]}
+    entry_points={"console_scripts": ["qdarkstyle=qdarkstyle.__main__:main"]},
+    install_requires=[
+        'watchdog',
```
Comment (dpizetta) [inline setup.py]: This dependency may should be optional since common users not need that.

File: script/process_qrc.py [hunk 3]
```diff
@@ -23,31 +23,76 @@
...
+    """QSS File observer."""
+
+    def __init__(self, parser_args):
+        """QSS File observer."""
+        super(QSSFileHandler, self).__init__()
+        self.args = parser_args
+
+    def on_modified(self, event):
+        """Handle file system events."""
+        if event.src_path.endswith('.qss'):
```
Comment (dpizetta) [inline script/process_qrc.py]: As mentioned the .qss file is not found in this  path, so the process_qrc is not running, but strangely  the files get their modification date updated.

File: setup.py (hunk 2 above)
Comment (dpizetta) [reply setup.py]: This dependency may should be optional since common users not need that.

File: script/process_qrc.py [hunk 4]
```diff
@@ -23,31 +23,74 @@
...
                         type=str,
                         help="Choose which one would be generated.")
+    parser.add_argument('--watch', '-w',
+                        action='store_true',
+                        help="Watch for file changes.")
 
     args = parser.parse_args(arguments)
 
+    if args.watch:
+        path = os.path.join(REPO_ROOT, 'qdarkstyle')
```
Comment (goanpeca) [inline script/process_qrc.py]: @dpizetta question 

is `/home/pizetta/workspace/qdark-combo/qdarkstyle/` the repo root or the module?

what does `path` print and what are the contents of that path?

File: script/process_qrc.py [hunk 5]
```diff
@@ -23,31 +23,74 @@
...
+
+class QSSFileHandler(FileSystemEventHandler):
+    """QSS File observer."""
+
+    def __init__(self, parser_args):
+        """QSS File observer."""
+        super(QSSFileHandler, self).__init__()
+        self.args = parser_args
+
+    def on_modified(self, event):
```
Comment (goanpeca) [inline script/process_qrc.py]: @dpizetta if you change `on _modified` to `on_any_event` still nothing?

File: script/process_qrc.py (hunk 4 above)
Comment (dpizetta) [reply script/process_qrc.py]: I tried something here, started with a fresh clone, and it worked! Running from root, script and qdarkstyle package folder.

Just answering: qdarkstyle is the path to package, the clone directory is qdark-combo (your other branch).

No problems anymore...sorry for that

File: script/process_qrc.py (hunk 5 above)
Comment (dpizetta) [reply script/process_qrc.py]: Solved in the last conversation :)

File: setup.py (hunk 2 above)
Comment (dpizetta) [reply setup.py]: What about add watchdog as optional dependencies[ (extra_requires](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies)) - as we will deliver compiled files for ordinary users (qdarkstyle as a dependency for another package), may those tools to generate the qss file should stay for developers? So, we keep only hard dependencies on install_requires in setup.py. (Thus, add qtsass to extra too) Agreed? @ColinDuquesnoy  @goanpeca

Comment (ColinDuquesnoy) [reply setup.py]: For now, I would add watchdog and qtsass to the extra_requires and helpdev to the normal dependencies. There is no problem adding a python dependency that is used in the end-user python package. Anything else should go in the dev dependencies.

Comment (dpizetta) [reply setup.py]: I've added watchdog to extras_require :) Also to run examples.

**Comment by goanpeca** [comment]
@dpizetta merged with master to include your changes :-)

**Comment by dpizetta** [comment]
Nice! tks

**Pull request merged**